In [1]:
C:\Users\sriva\OneDrive\Documents\Studies\SAIT\Python

SyntaxError: unexpected character after line continuation character (<ipython-input-1-bca9e0e9eb75>, line 1)

In [2]:
pwd

'C:\\Users\\sriva'

## Q1.How many rows and columns data has?
## Answer: (11183,11)

In [3]:
import numpy
import pandas

print(pandas.read_csv("well_data.csv"))
data=pandas.read_csv("well_data.csv")

# print data by calling it below
data

          DEPTH_MD   WELL        X_LOC      Y_LOC         GROUP  \
0       869.208000  W0913  437642.0625  6470974.0  NORDLAND GP.   
1       869.968000  W0913  437642.0625  6470974.0  NORDLAND GP.   
2       870.728000  W0913  437642.0625  6470974.0  NORDLAND GP.   
3       871.488000  W0913  437642.0313  6470974.0  NORDLAND GP.   
4       872.248000  W0913  437642.0313  6470974.0  NORDLAND GP.   
...            ...    ...          ...        ...           ...   
11178  4100.137436  W0204  482078.2500  6847164.5    DUNLIN GP.   
11179  4100.897436  W0204  482078.3125  6847164.5    DUNLIN GP.   
11180  4101.657436  W0204  482078.3750  6847164.5    DUNLIN GP.   
11181  4102.417436  W0204  482078.4375  6847164.5    DUNLIN GP.   
11182  4103.177436  W0204  482078.5000  6847164.5    DUNLIN GP.   

           FORMATION         DTC        PEF      NPHI      RHOB lithology_name  
0         Utsira Fm.  140.845428   1.453022  0.419031  2.076638          Shale  
1         Utsira Fm.  144.230087 

DEPTH_MD   WELL        X_LOC      Y_LOC         GROUP  \
0       869.208000  W0913  437642.0625  6470974.0  NORDLAND GP.   
1       869.968000  W0913  437642.0625  6470974.0  NORDLAND GP.   
2       870.728000  W0913  437642.0625  6470974.0  NORDLAND GP.   
3       871.488000  W0913  437642.0313  6470974.0  NORDLAND GP.   
4       872.248000  W0913  437642.0313  6470974.0  NORDLAND GP.   
...            ...    ...          ...        ...           ...   
11178  4100.137436  W0204  482078.2500  6847164.5    DUNLIN GP.   
11179  4100.897436  W0204  482078.3125  6847164.5    DUNLIN GP.   
11180  4101.657436  W0204  482078.3750  6847164.5    DUNLIN GP.   
11181  4102.417436  W0204  482078.4375  6847164.5    DUNLIN GP.   
11182  4103.177436  W0204  482078.5000  6847164.5    DUNLIN GP.   

           FORMATION         DTC        PEF      NPHI      RHOB lithology_name  
0         Utsira Fm.  140.845428   1.453022  0.419031  2.076638          Shale  
1         Utsira Fm.  144.230087   1.441285  0.425482  2.010360      Sandstone  
2         Utsira Fm.  145.840134   1.457139  0.436616  2.009847      Sandstone  
3         Utsira Fm.  145.741058   1.322686  0.514372  1.521392      Sandstone  
4         Utsira Fm.         NaN        NaN       NaN  1.609095      Sandstone  
...              ...         ...        ...       ...       ...            ...  
11178  Statfjord Fm.   81.683586   4.618176  0.191501  2.511264          Shale  
11179  Statfjord Fm.   81.818054   4.756323  0.314873  2.492342      Sandstone  
11180  Statfjord Fm.   81.847351   4.571058  0.340838  2.502546      Sandstone  
11181  Statfjord Fm.  101.491325   6.626430  0.414516  2.379578      Sandstone  
11182  Statfjord Fm.   88.571071  12.839459  0.355905  2.500079      Sandstone  

[11183 rows x 11 columns]

## Q2.What kind of data type is not available in dataset?
## Answer: int32

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11183 entries, 0 to 11182
Data columns (total 11 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   DEPTH_MD        11183 non-null  float64
 1   WELL            11183 non-null  object 
 2   X_LOC           11183 non-null  float64
 3   Y_LOC           11183 non-null  float64
 4   GROUP           11183 non-null  object 
 5   FORMATION       11172 non-null  object 
 6   DTC             11118 non-null  float64
 7   PEF             11130 non-null  float64
 8   NPHI            11120 non-null  float64
 9   RHOB            11143 non-null  float64
 10  lithology_name  11146 non-null  object 
dtypes: float64(7), object(4)
memory usage: 961.2+ KB


## Q3.How many null values do we have in dataset in total?
## Answer: 269

In [5]:
well_data = pandas.read_csv('well_data.csv')
print(well_data.isnull().sum())
well_data

DEPTH_MD           0
WELL               0
X_LOC              0
Y_LOC              0
GROUP              0
FORMATION         11
DTC               65
PEF               53
NPHI              63
RHOB              40
lithology_name    37
dtype: int64


DEPTH_MD   WELL        X_LOC      Y_LOC         GROUP  \
0       869.208000  W0913  437642.0625  6470974.0  NORDLAND GP.   
1       869.968000  W0913  437642.0625  6470974.0  NORDLAND GP.   
2       870.728000  W0913  437642.0625  6470974.0  NORDLAND GP.   
3       871.488000  W0913  437642.0313  6470974.0  NORDLAND GP.   
4       872.248000  W0913  437642.0313  6470974.0  NORDLAND GP.   
...            ...    ...          ...        ...           ...   
11178  4100.137436  W0204  482078.2500  6847164.5    DUNLIN GP.   
11179  4100.897436  W0204  482078.3125  6847164.5    DUNLIN GP.   
11180  4101.657436  W0204  482078.3750  6847164.5    DUNLIN GP.   
11181  4102.417436  W0204  482078.4375  6847164.5    DUNLIN GP.   
11182  4103.177436  W0204  482078.5000  6847164.5    DUNLIN GP.   

           FORMATION         DTC        PEF      NPHI      RHOB lithology_name  
0         Utsira Fm.  140.845428   1.453022  0.419031  2.076638          Shale  
1         Utsira Fm.  144.230087   1.441285  0.425482  2.010360      Sandstone  
2         Utsira Fm.  145.840134   1.457139  0.436616  2.009847      Sandstone  
3         Utsira Fm.  145.741058   1.322686  0.514372  1.521392      Sandstone  
4         Utsira Fm.         NaN        NaN       NaN  1.609095      Sandstone  
...              ...         ...        ...       ...       ...            ...  
11178  Statfjord Fm.   81.683586   4.618176  0.191501  2.511264          Shale  
11179  Statfjord Fm.   81.818054   4.756323  0.314873  2.492342      Sandstone  
11180  Statfjord Fm.   81.847351   4.571058  0.340838  2.502546      Sandstone  
11181  Statfjord Fm.  101.491325   6.626430  0.414516  2.379578      Sandstone  
11182  Statfjord Fm.   88.571071  12.839459  0.355905  2.500079      Sandstone  

[11183 rows x 11 columns]

## Q4.Delete all rows that these column values all missing: (DTC, PEF, NPHI, RHOB, lithology_name). What is the shape of dataset after deleting? Hint: this is very important step: DO NOT delete a row(s) that one, two, three, or four of those logs have null values. All 6 columns should be missing in that row.

## Answer: (11146, 11)





In [6]:
import pandas
missing_values = ["n/a", "na", "--"]
df = pandas.read_csv("well_data.csv", na_values = missing_values)

df = df.dropna(subset=['DTC', 'PEF', 'NPHI', 'RHOB', 'lithology_name'], how='all')
df

DEPTH_MD   WELL        X_LOC      Y_LOC         GROUP  \
0       869.208000  W0913  437642.0625  6470974.0  NORDLAND GP.   
1       869.968000  W0913  437642.0625  6470974.0  NORDLAND GP.   
2       870.728000  W0913  437642.0625  6470974.0  NORDLAND GP.   
3       871.488000  W0913  437642.0313  6470974.0  NORDLAND GP.   
4       872.248000  W0913  437642.0313  6470974.0  NORDLAND GP.   
...            ...    ...          ...        ...           ...   
11178  4100.137436  W0204  482078.2500  6847164.5    DUNLIN GP.   
11179  4100.897436  W0204  482078.3125  6847164.5    DUNLIN GP.   
11180  4101.657436  W0204  482078.3750  6847164.5    DUNLIN GP.   
11181  4102.417436  W0204  482078.4375  6847164.5    DUNLIN GP.   
11182  4103.177436  W0204  482078.5000  6847164.5    DUNLIN GP.   

           FORMATION         DTC        PEF      NPHI      RHOB lithology_name  
0         Utsira Fm.  140.845428   1.453022  0.419031  2.076638          Shale  
1         Utsira Fm.  144.230087   1.441285  0.425482  2.010360      Sandstone  
2         Utsira Fm.  145.840134   1.457139  0.436616  2.009847      Sandstone  
3         Utsira Fm.  145.741058   1.322686  0.514372  1.521392      Sandstone  
4         Utsira Fm.         NaN        NaN       NaN  1.609095      Sandstone  
...              ...         ...        ...       ...       ...            ...  
11178  Statfjord Fm.   81.683586   4.618176  0.191501  2.511264          Shale  
11179  Statfjord Fm.   81.818054   4.756323  0.314873  2.492342      Sandstone  
11180  Statfjord Fm.   81.847351   4.571058  0.340838  2.502546      Sandstone  
11181  Statfjord Fm.  101.491325   6.626430  0.414516  2.379578      Sandstone  
11182  Statfjord Fm.   88.571071  12.839459  0.355905  2.500079      Sandstone  

[11146 rows x 11 columns]

## Q.5.Filling missing values in FORMATION column using 'ffill' will leave still some null value in dataset. True or False?


## Answer: True
There are errors, as "Formation" is a string column and we can't calculate Mean for strings



In [ ]:
DTC_mean = df.['DTC'].mean()
PEF_mean = df.['PEF'].mean()
RHOB_mean = df.['RHOB'].mean()
NPHI_mean = df.['NPHI'].mean()

values = {'DTC': DTC_mean, 'PEF': PEF_mean, 'RHOB': RHOB_mean, 'NPHI': NPHI_mean}
df.fillna(value=values)

In [7]:
data['FORMATION'].mean()
print(data['FORMATION'])

TypeError: can only concatenate str (not "int") to str

## Q6.How many unique wells do we have in dataset?
## Answer: 5

In [8]:
unique_value = well_data.nunique()
print('Here is the count of unique values in each column:')
print(unique_value)

Here is the count of unique values in each column:
DEPTH_MD          10825
WELL                  5
X_LOC              3167
Y_LOC               148
GROUP                11
FORMATION            32
DTC               11088
PEF               11035
NPHI              11113
RHOB              11028
lithology_name        9
dtype: int64


# Q7.Count of unique "lithology_name".
## How many unique lithology_name is in dataset? Which one is the least frequent?
## Answer: 9 - Anhydrite

In [9]:
well_data['lithology_name'].unique()

array(['Shale', 'Sandstone', 'Sandstone/Shale', 'Limestone', 'Tuff',
       'Marl', 'Anhydrite', 'Dolomite', nan, 'Chalk'], dtype=object)

In [10]:
well_data_set=set(well_data)
no_of_unique_values=len(well_data_set)
print(no_of_unique_values)

11


In [11]:
df['lithology_name'].value_counts()


Shale              5149
Sandstone/Shale    2911
Sandstone          1691
Limestone           655
Marl                405
Chalk               217
Tuff                106
Dolomite              9
Anhydrite             3
Name: lithology_name, dtype: int64

## Q8.What is the x and y coordinate for each well? (compute mean value of X_loc, Y_loc for each well).
## Answer is 3rd option:  { W0204 :(482071.28, 6847165 ), W0913 :(437624.95, 6470980 ), W0915 :(436807.59, 6463004 ), W1005 :(448800.82, 6429624 ), W1105 :(463756.00, 6562935 )}

In [12]:
df = pandas.DataFrame(data)
df[['WELL','X_LOC','Y_LOC']]

WELL        X_LOC      Y_LOC
0      W0913  437642.0625  6470974.0
1      W0913  437642.0625  6470974.0
2      W0913  437642.0625  6470974.0
3      W0913  437642.0313  6470974.0
4      W0913  437642.0313  6470974.0
...      ...          ...        ...
11178  W0204  482078.2500  6847164.5
11179  W0204  482078.3125  6847164.5
11180  W0204  482078.3750  6847164.5
11181  W0204  482078.4375  6847164.5
11182  W0204  482078.5000  6847164.5

[11183 rows x 3 columns]

In [13]:
df.groupby('WELL').mean('X_LOC')

DEPTH_MD          X_LOC         Y_LOC         DTC       PEF  \
WELL                                                                    
W0204  2786.919642  482071.286610  6.847165e+06  113.650827  4.543725   
W0913  2070.304402  437624.954509  6.470980e+06  120.554527  3.658122   
W0915  2037.779972  436807.608537  6.463004e+06  121.885645  3.750773   
W1005  2817.220387  448798.324899  6.429616e+06   91.793687  7.167786   
W1105  1772.780719  463756.000000  6.562935e+06  123.989640  2.913153   

           NPHI      RHOB  
WELL                       
W0204  0.346227  2.325767  
W0913  0.411365  2.154323  
W0915  0.390958  2.157110  
W1005  0.249529  2.438222  
W1105  0.423095  2.118207

## Q9.What is the min, max, mean value of DTC in FORMATION ‘Utsira Fm.’ In whole dataset?
## Answer: min: 30.35, max: 652.93, mean:147.74


In [14]:
Utsira_fm=data[data["FORMATION"]=='Utsira Fm.']
Utsira_fm.describe()

DEPTH_MD          X_LOC         Y_LOC          DTC          PEF  \
count  3864.000000    3864.000000  3.864000e+03  3844.000000  3839.000000   
mean   1530.834676  444389.234247  6.527747e+06   147.784726     2.730680   
std     344.829585   16403.922213  1.390378e+05    18.884695     1.361976   
min     869.208000  436809.406300  6.462990e+06    30.353481     1.011461   
25%    1244.344000  436813.054700  6.462994e+06   143.539867     2.229955   
50%    1564.532000  437631.484400  6.470976e+06   147.825432     2.616765   
75%    1812.001077  437640.843800  6.470979e+06   153.633778     2.905853   
max    2197.536000  482079.500000  6.847170e+06   652.939621    32.089067   

              NPHI         RHOB  
count  3843.000000  3848.000000  
mean      0.489111     1.979334  
std       0.047974     0.108038  
min       0.274227     1.500153  
25%       0.453945     1.936907  
50%       0.485386     1.998336  
75%       0.521660     2.035887  
max       0.697367     2.458299

## Q10. What is the min, max, mean value of NPHI in FORMATION ‘Utsira Fm.’ In the WELL ‘ W0204'?
## Answer: min: 0.2742,   max: 0.4956,  mean: 0.4463

In [15]:
Utsira_form_well = data[(data["FORMATION"]=='Utsira Fm.') & (data["WELL"]=='W0204')]
Utsira_form_well.describe()

DEPTH_MD          X_LOC         Y_LOC         DTC         PEF  \
count   615.000000     615.000000  6.150000e+02  615.000000  615.000000   
mean   1706.466646  482078.212728  6.847169e+06  150.487827    2.664577   
std     138.318000       0.782771  3.199451e-01   13.568611    0.093400   
min    1470.385436  482076.937500  6.847168e+06  102.199493    2.535711   
25%    1587.045436  482077.531300  6.847168e+06  143.071007    2.620169   
50%    1703.705436  482078.218800  6.847169e+06  153.198105    2.651739   
75%    1829.029436  482078.906300  6.847169e+06  159.467361    2.697839   
max    1945.689436  482079.500000  6.847170e+06  179.712250    3.497786   

             NPHI        RHOB  
count  615.000000  615.000000  
mean     0.446378    2.021581  
std      0.034979    0.096589  
min      0.274227    1.813545  
25%      0.428375    1.957660  
50%      0.457068    2.002288  
75%      0.470368    2.074379  
max      0.495601    2.365326

## 11.Create a function to calculate porosity from RHOB using this equation as fraction. How many arguments can have default value for this function?
## Answer: Can have 1 default value


In [16]:
def densityPorosity(rb, rf, rm):
    '''
    rb = bulk density (RHOB)
    rf = fluid density (default value of 1)
    rm = matrix density (limeston: 2.71 g/cc)
    '''
    por = (rm-rb)/(rm-rf)
    ron_por = round(por, 2)
    return ron_por


## Q12.Create new column called ‘Den_por’ and apply the function to calculate porosity in that column. Can 'apply' function receive another function as an argument?
## Answer: Yes

In [17]:
def densityPorosity(rb, rf, rm):
    '''
    rb = bulk density (RHOB)
    rf = fluid density (default value of 1)
    rm = matrix density (limeston: 2.71 g/cc)
    '''
    por = (rm-rb)/(rm-rf)
    data['Den_por'] = round(por, 2)
    return data['Den_por']

densityPorosity(data['RHOB'],1, 2.71)

0        0.37
1        0.41
2        0.41
3        0.70
4        0.64
         ... 
11178    0.12
11179    0.13
11180    0.12
11181    0.19
11182    0.12
Name: Den_por, Length: 11183, dtype: float64

In [18]:
data['Den_por'].head()

0    0.37
1    0.41
2    0.41
3    0.70
4    0.64
Name: Den_por, dtype: float64

In [19]:
print(data['Den_por'])

0        0.37
1        0.41
2        0.41
3        0.70
4        0.64
         ... 
11178    0.12
11179    0.13
11180    0.12
11181    0.19
11182    0.12
Name: Den_por, Length: 11183, dtype: float64


In [20]:
df=pandas.DataFrame(data)
df[['lithology_name','Den_por','RHOB']]

lithology_name  Den_por      RHOB
0              Shale     0.37  2.076638
1          Sandstone     0.41  2.010360
2          Sandstone     0.41  2.009847
3          Sandstone     0.70  1.521392
4          Sandstone     0.64  1.609095
...              ...      ...       ...
11178          Shale     0.12  2.511264
11179      Sandstone     0.13  2.492342
11180      Sandstone     0.12  2.502546
11181      Sandstone     0.19  2.379578
11182      Sandstone     0.12  2.500079

[11183 rows x 3 columns]

## Q13.Calculate ‘PHIT’ (total porosity) column using lambda function. How many parameters lambda function will have for this calculation?
## Answer: 2

In [21]:
df['PHIT']= df.apply(lambda row: (row.NPHI + row.Den_por)/2, axis= 1)
df.head()


DEPTH_MD   WELL        X_LOC      Y_LOC         GROUP   FORMATION  \
0   869.208  W0913  437642.0625  6470974.0  NORDLAND GP.  Utsira Fm.   
1   869.968  W0913  437642.0625  6470974.0  NORDLAND GP.  Utsira Fm.   
2   870.728  W0913  437642.0625  6470974.0  NORDLAND GP.  Utsira Fm.   
3   871.488  W0913  437642.0313  6470974.0  NORDLAND GP.  Utsira Fm.   
4   872.248  W0913  437642.0313  6470974.0  NORDLAND GP.  Utsira Fm.   

          DTC       PEF      NPHI      RHOB lithology_name  Den_por      PHIT  
0  140.845428  1.453022  0.419031  2.076638          Shale     0.37  0.394515  
1  144.230087  1.441285  0.425482  2.010360      Sandstone     0.41  0.417741  
2  145.840134  1.457139  0.436616  2.009847      Sandstone     0.41  0.423308  
3  145.741058  1.322686  0.514372  1.521392      Sandstone     0.70  0.607186  
4         NaN       NaN       NaN  1.609095      Sandstone     0.64       NaN

## Q14.What is the min, max, mean value of PHIT in FORMATION ‘Lista Fm.’ In the WELL ‘’ W0913?
## Answer: min: 0.3226, max: 0.6602, mean: 0.5053

In [22]:
PHIT_table=pandas.DataFrame(df)

In [23]:
Lista_form_well = data[(data["FORMATION"]=='Lista Fm.') & (data["WELL"]=='W0913')]
Lista_form_well.describe()

DEPTH_MD          X_LOC         Y_LOC         DTC        PEF  \
count    85.000000      85.000000  8.500000e+01   85.000000  85.000000   
mean   2362.456000  437619.127962  6.470981e+06  129.432840   5.835869   
std      18.757789       0.485838  2.464027e-01   11.880257   1.197297   
min    2330.536000  437618.312500  6.470981e+06  107.319557   3.183859   
25%    2346.496000  437618.718800  6.470981e+06  120.320854   5.009374   
50%    2362.456000  437619.125000  6.470982e+06  131.384872   5.944208   
75%    2378.416000  437619.531300  6.470982e+06  136.184296   6.749939   
max    2394.376000  437619.968800  6.470982e+06  161.097290   8.734501   

            NPHI       RHOB    Den_por  
count  85.000000  85.000000  85.000000  
mean    0.521298   1.873012   0.489529  
std     0.072743   0.168242   0.098427  
min     0.323443   1.543206   0.300000  
25%     0.478382   1.738321   0.420000  
50%     0.526964   1.862923   0.500000  
75%     0.575742   1.988331   0.570000  
max     0.656211   2.202501   0.680000

## Q15.Find the outliers for DTC, PEF, and NPHI, quantitatively. How many outliers are there for each feature? Which Well(s) has the least outliers?
## Answer: DTC: 5, PEF: 4, NPHI: 11
## W0913 and W1105 both have the least outliers (4 outliers each)

In [26]:
data.describe()

DEPTH_MD          X_LOC         Y_LOC           DTC           PEF  \
count  11183.000000   11183.000000  1.118300e+04  11118.000000  11130.000000   
mean    2291.844464  453985.717644  6.590754e+06    117.721368      4.058413   
std      764.484477   20197.940783  1.732961e+05     31.032229      2.158162   
min      869.208000  436790.843800  6.429286e+06     30.353481      1.011461   
25%     1683.936718  436813.593800  6.463013e+06     91.138182      2.632901   
50%     2241.008000  437640.687500  6.470982e+06    117.697723      3.319412   
75%     2865.718193  482061.171900  6.847156e+06    145.932335      4.878673   
max     4103.177436  482080.218800  6.847174e+06    652.939621     32.089067   

               NPHI          RHOB       Den_por  
count  11120.000000  11143.000000  11143.000000  
mean       0.379076      2.218300      0.287618  
std        0.123957      0.247333      0.144732  
min       -0.235535      1.500153     -0.120000  
25%        0.291939      2.009836      0.150000  
50%        0.393674      2.205459      0.300000  
75%        0.479975      2.453044      0.410000  
max        0.699346      2.910462      0.710000

In [33]:
df_dtc = df.loc[:,['WELL','DTC']][(df['DTC'] > 250) | (df['DTC'] < 25) ]
df_dtc

WELL         DTC
3964  W0915  554.183075
3965  W0915  354.458328
3966  W0915  652.939621
3970  W0915  558.033463
3971  W0915  449.934052

In [34]:
df_pef = df.loc[:,['WELL','PEF']][(df['PEF'] > 15) | (df['PEF'] < 1) ]
df_pef

WELL        PEF
3963  W0915  22.067261
3964  W0915  18.210827
3965  W0915  32.089067
3966  W0915  21.103240

In [35]:
df_nphi = df.loc[:,['WELL','NPHI']][(df['NPHI'] > 0.65) | (df['NPHI'] < -0.15) ]
df_nphi

WELL      NPHI
357   W0913  0.690551
367   W0913  0.653731
1601  W0913  0.697367
1998  W0913  0.656211
5565  W0915 -0.172432
5580  W0915 -0.232258
5591  W0915 -0.235535
6733  W1105  0.654904
6734  W1105  0.684932
6735  W1105  0.699346
6736  W1105  0.680164

## Q16.Which Group has the most various FORMATION in dataset?
## Answer: Shetland Gp. has 8 different formations - the most various formations

In [36]:
Group_mostVarious_Formation=df.groupby(['GROUP','FORMATION']).FORMATION.count()
print(Group_mostVarious_Formation)

GROUP             FORMATION            
CROMER KNOLL GP.  Aasgard Fm.               175
                  Blodoeks Fm.               42
                  Lista Fm.                  20
                  Roedby Fm.                336
                  Sola Fm.                   67
                  Tuxen Fm.                  43
DUNLIN GP.        Amundsen Fm.              120
                  Burton Fm.                 51
                  Cook Fm.                  145
                  Drake Fm.                 269
                  Statfjord Fm.              53
HEGRE GP.         Skagerrak Fm.             495
HORDALAND GP.     Frigg Fm.                 237
                  Skade Fm.                 359
                  Utsira Fm.               3116
NORDLAND GP.      Utsira Fm.                748
ROGALAND GP.      Balder Fm.                221
                  Heimdal Fm.               159
                  Intra Balder Fm. Sst.       4
                  Lista Fm.                 406


In [37]:
df01 = df.groupby(['GROUP']).FORMATION.nunique()
df01.head

<bound method NDFrame.head of GROUP
CROMER KNOLL GP.    6
DUNLIN GP.          5
HEGRE GP.           1
HORDALAND GP.       3
NORDLAND GP.        1
ROGALAND GP.        5
SHETLAND GP.        8
TYNE GP.            1
VESTLAND GP.        4
VIKING GP.          3
ZECHSTEIN GP.       1
Name: FORMATION, dtype: int64>